File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/connection/testing/ConnectionTester.java
package com.lasthopesoftware.bluewater.client.connection.testing;

import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider;
import com.lasthopesoftware.bluewater.shared.StandardRequest;
import com.namehillsoftware.handoff.promises.Promise;
import okhttp3.Response;
import okhttp3.ResponseBody;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;

public class ConnectionTester implements TestConnections {

	private static final Logger mLogger = LoggerFactory.getLogger(ConnectionTester.class);

	@Override
	public Promise<Boolean> promiseIsConnectionPossible(IConnectionProvider connectionProvider) {
		return connectionProvider.promiseResponse("Alive").then(this::doTestSynchronously, e -> false);
	}

	private boolean doTestSynchronously(Response response) {
		final ResponseBody body = response.body();
		if (body == null) return false;

		try (final InputStream is = body.byteStream()) {
			final StandardRequest responseDao = StandardRequest.fromInputStream(is);

			return responseDao != null && responseDao.isStatus();
		} catch (IOException e) {
			mLogger.error("Error closing connection, device failure?", e);
		} catch (IllegalArgumentException e) {
			mLogger.warn("Illegal argument passed in", e);
		} finally {
			body.close();
		}

		return false;
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/connection/testing/ConnectionTester.kt
package com.lasthopesoftware.bluewater.client.connection.testing

import com.lasthopesoftware.bluewater.client.connection.ProvideConnections
import com.lasthopesoftware.bluewater.shared.NonStandardResponseException
import com.lasthopesoftware.bluewater.shared.StandardResponse.Companion.toStandardResponse
import com.lasthopesoftware.bluewater.shared.lazyLogger
import com.namehillsoftware.handoff.cancellation.CancellationSignal
import com.namehillsoftware.handoff.promises.Promise
import com.namehillsoftware.handoff.promises.propagation.CancellationProxy
import okhttp3.Response
import org.jsoup.Jsoup
import org.jsoup.parser.Parser
import java.io.IOException

object ConnectionTester : TestConnections {
	private val logger by lazyLogger<ConnectionTester>()

	override fun promiseIsConnectionPossible(connectionProvider: ProvideConnections): Promise<Boolean> =
		ConnectionPossiblePromise(connectionProvider)

	private class ConnectionPossiblePromise(connectionProvider: ProvideConnections) : Promise<Boolean>() {
		init {
			val cancellationProxy = CancellationProxy()
			awaitCancellation(cancellationProxy)

			connectionProvider
				.promiseResponse("Alive")
				.also(cancellationProxy::doCancel)
				.then(
					{ it, cp -> resolve(testResponse(it, cp)) },
					{ e, _ ->
						logger.error("Error checking connection at URL {}.", connectionProvider.urlProvider.baseUrl, e)
						resolve(false)
					}
				)
				.also(cancellationProxy::doCancel)
		}
	}

	private fun testResponse(response: Response, cancellationSignal: CancellationSignal): Boolean {
		response.body.use { body ->
			if (cancellationSignal.isCancelled) return false

			try {
				return body.string().let { Jsoup.parse(it, Parser.xmlParser()) }.toStandardResponse().isStatusOk
			} catch (e: NonStandardResponseException) {
				logger.warn("Non standard response received.", e)
			} catch (e: IOException) {
				logger.error("Error closing connection, device failure?", e)
			} catch (e: IllegalArgumentException) {
				logger.warn("Illegal argument passed in", e)
			} catch (t: Throwable) {
				logger.error("Unexpected error parsing response.", t)
			}
		}
		return false
	}
}
